### PR TITLE
DEV-13 hashing database content

### PR DESCRIPTION
**Pull request by daroczig**

### CODE REVIEW

File: src/sc_crawler/cli.py [hunk 1]
```diff
@@ -10,8 +11,20 @@
 from typing_extensions import Annotated
 
 from . import vendors as vendors_module
+from .hashing import get_rows, get_table_name, hashrows
 from .logger import logger
-from .schemas import Vendor
+from .schemas import (
```
Comment (daroczig) [inline src/sc_crawler/cli.py:16]: might better set up `__all__` in schemas?

**Comment by daroczig** [comment]
Example runs:

```sh
$ sc-crawler hash --connection-string sqlite:///sc-data-all.db
4476ce09a367d55e3f048820e0bde284c5c1405a
```

```sh
$ sc-crawler hash --connection-string sqlite:///sc-data-all.db --level table
{"addon_storage": "bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f", "addon_traffic": "bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f", "complianceframework": "10054e710ca05b68e47964b0e26a60e9696db431", "country": "59a358b4ea892805e1c8d9eb0683d6c2c62caa87", "datacenter": "247500b1a34a118ed7df89f189500b12493e9e3f", "price": "33ddcc1113e423382c255d34a6b6ecf9daedecff", "server": "81623bfeeabb34dbe67e757f9d8981ec3c447905", "vendor": "733e657005dd427e8f7e98a7399cdb4ae43f99b1", "vendor_compliance_link": "46833fa3960e757de3250a687c18124d44cedb66", "zone": "c1c06067434c71bed75cb56efb272e1a51ee3c8f"}
```

```sh
$ sc-crawler hash --connection-string sqlite:///sc-data-all.db --level row
{"addon_storage": {}, "addon_traffic": {}, "complianceframework": {"('hipaa',)": "f2b6ccef77be3cda1177a60fd2dee0b77c66d3f7", "('soc2t2',)": "904979fbaab0703ac1c4f3b2c28aa21431e64bcd"}, "country": {"('AE',)": "7dc85578d86d37d40bd0e4653aa535
32c2441705", "('AU',)": "81aa4385a1ec693b9824d53e1ee5cd5ee4bd90dc", "('BH',)": "7dc85578d86d37d40bd0e4653aa53532c2441705", "('BR',)": "e2557f83ede08513ab755880b14444d95e0459ce", "('CA',)": "e2bc5d42e30e015ee3390a5c6666c978c82d5c8b", "('CH'
,)": "2cbe94de7be7dd7b1ec688a9f1297cd631e30594", "('CN',)": "7dc85578d86d37d40bd0e4653aa53532c2441705", ...

File: src/sc_crawler/cli.py [hunk 2]
```diff
@@ -50,6 +57,39 @@ def metadata_dump(sql, *_args, **_kwargs):
...
+        }
+        print(dumps(table_hashes, sort_keys=True))
+        raise typer.Exit()
+
+    json_dump = dumps(table_hashes, sort_keys=True)
+    if level == HashLevels.ROW:
+        print(json_dump)
+
+    if level == HashLevels.DATABASE:
+        print(sha1(json_dump.encode()).hexdigest())
```
Comment (bra-fsn) [inline src/sc_crawler/cli.py:90]: Provide a consistent output type here maybe? Either JSON or a formatted text output (not sure about the `ROW` type, but for the others, the latter might be the best?).

**Review by bra-fsn — APPROVED** [review]
BTW, my editor shows a lot of errors (turning all of these into green sometimes take more time than writing the actual code, at least for me :).
mypy gives these:
```
src/sc_crawler/schemas.py:30: error: Incompatible return value type (got "str | Callable[..., str]", expected "str")  [return-value]
src/sc_crawler/schemas.py:33: error: Bracketed expression "[...]" is not valid as a type  [valid-type]
src/sc_crawler/schemas.py:34: error: Item "None" of "Any | None" has no attribute "primary_key"  [union-attr]
src/sc_crawler/schemas.py:142: error: Incompatible types in assignment (expression has type "str", variable has type "Status")  [assignment]
src/sc_crawler/schemas.py:214: error: Incompatible types in assignment (expression has type "str", variable has type "Status")  [assignment]
src/sc_crawler/schemas.py:227: error: Incompatible types in assignment (expression has type "str", variable has type "Status")  [assignment]
src/sc_crawler/schemas.py:256: error: Incompatible types in assignment (expression has type "str", variable has type "Status")  [assignment]
src/sc_crawler/schemas.py:276: error: Incompatible types in assignment (expression has type "str", variable has type "Status")  [assignment]
src/sc_crawler/schemas.py:447: error: Incompatible types in assignment (expression has type "str", variable has type "Allocation")  [assignment]
src/sc_crawler/vendors/aws.py:7: error: Skipping analyzing "boto3": module is installed, but missing library stubs or py.typed marker  [import-untyped]
src/sc_crawler/vendors/aws.py:7: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
src/sc_crawler/cli.py:22: error: Second argument of Enum() must be string, tuple, list or dict literal for mypy to determine Enum members  [misc]
src/sc_crawler/cli.py:33: error: Second argument of Enum() must be string, tuple, list or dict literal for mypy to determine Enum members  [misc]
src/sc_crawler/cli.py:37: error: Second argument of Enum() must be string, tuple, list or dict literal for mypy to determine Enum members  [misc]
src/sc_crawler/cli.py:70: error: Incompatible default for argument "level" (default has type "str", argument has type "HashLevels")  [assignment]
src/sc_crawler/cli.py:108: error: Incompatible default for argument "log_level" (default has type "str", argument has type "LogLevels")  [assignment]
Found 15 errors in 3 files (checked 10 source files)
```

File: src/sc_crawler/cli.py (hunk 2 above)
Comment (daroczig) [reply src/sc_crawler/cli.py:90]: good point :+1: 

what about adding a `format` option, defaults to JSON, but supporting YAML as well. that should be good for DB and TABLE levels, and good enough for ROW as well

Comment (bra-fsn) [reply src/sc_crawler/cli.py:90]: > good point 👍
> 
> what about adding a `format` option, defaults to JSON, but supporting YAML as well? that should be good for DB and TABLE levels, and good enough for ROW as well

In FreeBSD a lot of tools use [libxo](https://github.com/Juniper/libxo), which provides a similar capability for generic CLI tools. Nice idea, although I'm not sure if anybody will ever use it. :)

Comment (daroczig) [reply src/sc_crawler/cli.py:90]: the only reason to have two output types in my opinion was to provide a way to return a single hash without curly braces :) I think I'll just drop the `--level` support and keep it at the DB level, and for any other levels, python helper can be used

**Pull request merged**